File: src/main/java/guru/xgm/jass/codeInspection/JassOperatorCallBehaviorInspection.java
package guru.xgm.jass.codeInspection;

import com.intellij.codeInspection.*;
import com.intellij.codeInspection.options.OptPane;
import com.intellij.lang.ASTNode;
import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiElementVisitor;
import com.intellij.psi.PsiParserFacade;
import com.intellij.psi.PsiWhiteSpace;
import guru.xgm.jass.psi.*;
import org.jetbrains.annotations.NotNull;

import static com.intellij.codeInspection.options.OptPane.*;

final class JassOperatorCallBehaviorInspection extends LocalInspectionTool implements CleanupLocalInspectionTool {
    private final MissingQuickFix missingQuickFix = new MissingQuickFix();
    private final RedundantQuickFix redundantQuickFix = new RedundantQuickFix();

    enum ErrorWhen {
        MISSING,
        REDUNDANT,
    }

    /**
     * @noinspection PublicField
     */
    public ErrorWhen errorWhen = ErrorWhen.MISSING;

    @Override
    public @NotNull OptPane getOptionsPane() {
        return pane(dropdown(
                "errorWhen",
                JassInspectionBundle.message("operator.call.dropdown.label"),
                option(ErrorWhen.MISSING, JassInspectionBundle.message("operator.call.dropdown.option.missing")),
                option(ErrorWhen.REDUNDANT, JassInspectionBundle.message("operator.call.dropdown.option.redundant"))
        ));

    }

    @NotNull
    @Override
    public PsiElementVisitor buildVisitor(@NotNull final ProblemsHolder holder, boolean isOnTheFly) {

        return new JassVisitor() {
            @Override
            public void visitCallStmt(@NotNull JassCallStmt callStmt) {
                super.visitPsiElement(callStmt);

                final JassFunCall funcCall = callStmt.getFunCall();
                final PsiElement funcCallName = funcCall.getId();

                final ASTNode callToken = callStmt.getNode().findChildByType(JassTypes.CALL);
                if (callToken == null) {
                    if (errorWhen == ErrorWhen.MISSING) {
                        holder.registerProblem(
                                funcCallName,
                                JassInspectionBundle.message("operator.call.missing.descriptor"),
                                missingQuickFix
                        );
                    }
                } else {
                    if (errorWhen == ErrorWhen.REDUNDANT) {
                        holder.registerProblem(
                                funcCallName,
                                JassInspectionBundle.message("operator.call.redundant.descriptor"),
                                redundantQuickFix
                        );
                    }
                }
            }
        };
    }

    private static class MissingQuickFix implements LocalQuickFix {
        @NotNull
        @Override
        public String getName() {
            return JassInspectionBundle.message("operator.call.missing.quickfix");
        }

        @NotNull
        public String getFamilyName() {
            return getName();
        }

        public void applyFix(@NotNull Project project, @NotNull ProblemDescriptor descriptor) {
            final PsiElement funcCallName = descriptor.getPsiElement();
            final JassFunCall funcCall = (JassFunCall) funcCallName.getParent();
            final JassCallStmt callStmt = (JassCallStmt) funcCall.getParent();

            final JassCallStmt callStmtClone = (JassCallStmt) callStmt.copy();
            final JassFunCall funcCallCopy = callStmtClone.getFunCall();

            callStmtClone.addBefore(JassElementFactory.createToken(project, "call"), funcCallCopy);
            callStmtClone.addBefore(PsiParserFacade.getInstance(project).createWhiteSpaceFromText(" "), funcCallCopy);

            callStmt.replace(JassElementFactory.recreateCallStmt(project, callStmtClone));
        }

    }

    private static class RedundantQuickFix implements LocalQuickFix {

        @NotNull
        @Override
        public String getName() {
            return JassInspectionBundle.message("operator.call.redundant.quickfix");
        }

        @NotNull
        public String getFamilyName() {
            return getName();
        }

        public void applyFix(@NotNull Project project, @NotNull ProblemDescriptor descriptor) {
            final PsiElement funcCallName = descriptor.getPsiElement();
            final JassFunCall funcCall = (JassFunCall) funcCallName.getParent();
            final PsiElement ws = funcCall.getPrevSibling();
            final PsiElement call = ws.getPrevSibling();
            if (!(ws instanceof PsiWhiteSpace) || call.getNode().getElementType() != JassTypes.CALL) {
                return;
            }
            ws.delete();
            call.delete();
        }
    }
}

File: src/main/kotlin/guru/xgm/language/jass/codeInspection/JassOperatorCallBehaviorInspection.kt
package guru.xgm.language.jass.codeInspection

import com.intellij.codeInspection.*
import com.intellij.codeInspection.options.OptPane
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiElementVisitor
import com.intellij.psi.PsiParserFacade
import com.intellij.psi.PsiWhiteSpace
import guru.xgm.language.jass.codeInspection.JassInspectionBundle.message
import guru.xgm.language.jass.psi.JassCallStmt
import guru.xgm.language.jass.psi.JassElementFactory.createToken
import guru.xgm.language.jass.psi.JassElementFactory.recreateCallStmt
import guru.xgm.language.jass.psi.JassFunCall
import guru.xgm.language.jass.psi.JassTypes
import guru.xgm.language.jass.psi.JassVisitor

internal class JassOperatorCallBehaviorInspection : LocalInspectionTool(), CleanupLocalInspectionTool {
    private val missingQuickFix = MissingQuickFix()
    private val redundantQuickFix = RedundantQuickFix()

    internal enum class ErrorWhen {
        MISSING,
        REDUNDANT,
    }

    var errorWhen: ErrorWhen = ErrorWhen.MISSING

    override fun getOptionsPane(): OptPane {
        return OptPane.pane(
            OptPane.dropdown(
                "errorWhen",
                message("operator.call.dropdown.label"),
                OptPane.option(ErrorWhen.MISSING, message("operator.call.dropdown.option.missing")),
                OptPane.option(ErrorWhen.REDUNDANT, message("operator.call.dropdown.option.redundant"))
            )
        )
    }

    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor {
        return object : JassVisitor() {
            override fun visitCallStmt(callStmt: JassCallStmt) {
                super.visitPsiElement(callStmt)

                val funcCall = callStmt.funCall
                val funcCallName = funcCall.id

                val callToken = callStmt.node.findChildByType(JassTypes.CALL)
                if (callToken == null) {
                    if (errorWhen == ErrorWhen.MISSING) {
                        holder.registerProblem(
                            funcCallName,
                            message("operator.call.missing.descriptor"),
                            missingQuickFix
                        )
                    }
                } else {
                    if (errorWhen == ErrorWhen.REDUNDANT) {
                        holder.registerProblem(
                            funcCallName,
                            message("operator.call.redundant.descriptor"),
                            redundantQuickFix
                        )
                    }
                }
            }
        }
    }

    private class MissingQuickFix : LocalQuickFix {
        override fun getName(): String {
            return message("operator.call.missing.quickfix")
        }

        override fun getFamilyName(): String {
            return name
        }

        override fun applyFix(project: Project, descriptor: ProblemDescriptor) {
            val funcCallName = descriptor.psiElement
            val funcCall = funcCallName.parent as JassFunCall
            val callStmt = funcCall.parent as JassCallStmt

            val callStmtClone = callStmt.copy() as JassCallStmt
            val funcCallCopy = callStmtClone.funCall

            callStmtClone.addBefore(createToken(project, "call"), funcCallCopy)
            callStmtClone.addBefore(PsiParserFacade.getInstance(project).createWhiteSpaceFromText(" "), funcCallCopy)

            callStmt.replace(recreateCallStmt(project, callStmtClone)!!)
        }
    }

    private class RedundantQuickFix : LocalQuickFix {
        override fun getName(): String {
            return message("operator.call.redundant.quickfix")
        }

        override fun getFamilyName(): String {
            return name
        }

        override fun applyFix(project: Project, descriptor: ProblemDescriptor) {
            val funcCallName = descriptor.psiElement
            val funcCall = funcCallName.parent as JassFunCall
            val ws = funcCall.prevSibling
            val call = ws.prevSibling
            if (ws !is PsiWhiteSpace || call.node.elementType !== JassTypes.CALL) return
            ws.delete()
            call.delete()
        }
    }
}
